torch/nn/utils/parametrizations.py

```diff
@@ -0,0 +1,181 @@
+        #     which is a new module object created on the fly. And each replica
+        #     runs its own spectral norm power iteration. So simply assigning
+        #     the updated vectors to the module this function runs on will cause
+        #     the update to be lost forever. And the next time the parallelized
+        #     module is replicated, the same randomly initialized vectors are
+        #     broadcast and used!
+        #
+        #     Therefore, to make the change propagate back, we rely on two
+        #     important behaviors (also enforced via tests):
+        #       1. `DataParallel` doesn't clone storage if the broadcast tensor
+        #          is already on correct device; and it makes sure that the
+        #          parallelized module is already on `device[0]`.
+        #       2. If the out tensor in `out=` kwarg has correct shape, it will
+        #          just fill in the values.
+        #     Therefore, since the same power iteration is performed on all
+        #     devices, simply updating the tensors in-place will make sure that
+        #     the module replica on `device[0]` will update the _u vector on the
+        #     parallized module (by shared storage).
+        #
+        #    However, after we update `u` and `v` in-place, we need to **clone**
+        #    them before using them to normalize the weight. This is to support
+        #    backproping through two forward passes, e.g., the common pattern in
+        #    GAN training: loss = D(real) - D(fake). Otherwise, engine will
+        #    complain that variables needed to do backward for the first forward
+        #    (i.e., the `u` and `v` vectors) are changed in the second forward.
+        weight_mat = self._reshape_weight_to_matrix(weight)
+
+        if self.training:
+            self.buffers_initialized = True
+            with torch.no_grad():
+                for _ in range(self.n_power_iterations):
+                    # Spectral norm of weight equals to `u^T W v`, where `u` and `v`
+                    # are the first left and right singular vectors.
+                    # This power iteration produces approximations of `u` and `v`.
+                    self.v = F.normalize(torch.mv(weight_mat.t(), self.u),  # type: ignore[has-type]
+                                         dim=0, eps=self.eps, out=self.v)   # type: ignore[has-type]
+                    self.u = F.normalize(torch.mv(weight_mat, self.v),
+                                         dim=0, eps=self.eps, out=self.u)   # type: ignore[has-type]
+                # See above on why we need to clone
+                self.u = self.u.clone(memory_format=torch.contiguous_format)
+                self.v = self.v.clone(memory_format=torch.contiguous_format)
+        else:
+            # https://github.com/pytorch/pytorch/issues/51800
+            if not self.buffers_initialized:
+                warnings.warn("SpectralNorm's `forward` called for the first time, but "
```

Comment:
Why do you need this? You are initializing v and u based on the given weight in the init no?

Reply:
I thought about this the other day and I forgot to leave a comment.
We have access to the `weight` during `__init__`, so I believe it'd be good to just initialize `u` and `v` correctly in the `__init__` and not have this else nor the `buffers_initialized` attribute.

Reply:
Fixed.